### ISSUE TITLE
Iterative Phases Skipped in MLE-STAR Workflow (Refinement, Ensemble, Validation)

### ISSUE THREAD

**Issue by Htlin-Lin** [Htlin-Lin/ClaudeFlow-MLEstar#1]
# Iterative Phases Skipped in MLE-STAR Workflow (Refinement, Ensemble, Validation)

## Context
- Command executed (in another project folder as per MLE-STAR guide):
  - `claude-flow automation mle-star --dataset "data/data_train.csv" --target MONTH_QTY --output "./outputs/" --name "Time Series Forecasting" --claude`
- Expected behavior: Run tasks per `src/cli/simple-commands/templates/mle-star-workflow.json` with agents and iterative phases (ablation → targeted refinement → ensemble → validation/deployment), driven by MLE-STAR methodology described in `wiki/MLE-STAR-Workflow.md`.
- Actual behavior: Early phases (env init, dataset analysis, web search, foundation build) show minutes of agent activity, files produced, and detailed logs. Later phases (ablation analysis, iterative refinement, ensemble, validation) complete in seconds (≤10s) with minimal or no outputs, indicating tasks were skipped or short-circuited.
- Evidence: `mle-star-1124.log` shows quick "COMPLETED" statuses for refinement-related tasks without agent work or outputs.

## Problem Statement
The workflow engine does not perform the iterative refinement phases (ablation analysis, targeted refinement, ensemble creation, and comprehensive validation) despite being defined in `mle-star-workflow.json`. Instead, it marks tasks as "COMPLETED" quickly with no substantive execution, suggesting dependency resolution, input binding, or execution gating is incorrectly bypassing these phases.

## Goals
- Ensure iterative phases run with real agent execution, produce artifacts/logs, and consume prior outputs via stream-chaining.
- Enforce task dependencies and input resolution so downstream phases receive required data structures.
- Add defensive checks so tasks cannot auto-complete when inputs are missing or empty.
- Provide observability (logs/metrics) to confirm iteration counts and per-iteration actions.

## Non-Goals
- Investigate early-phase "Failed: Process exited" causes. The early phases did, in practice, run sufficiently for this issue’s scope.
- Change the core methodology or reduce phases; focus on correctness of orchestration.

## Hypotheses (Root-Cause Analysis)
- Input Binding Missing/Incorrect:
  - Tasks reference `${...}` placeholders (e.g., `${ablation_analysis.output.highest_impact_component}`) that may resolve to `undefined/null` or empty objects when upstream outputs are not saved or are named differently.
  - If the runner treats missing inputs as acceptable, the task may be considered trivially complete.
- Dependency Enforcement Weakness:
  - The engine may mark tasks ready even if upstream tasks are failed or partial, due to `failurePolicy: continue` or lax gating.
  - Dependencies exist in both `tasks[].depends` and top-level `dependencies`; any mismatch or duplication could result in inconsistent checks.
- Iteration Controls Not Applied:
  - `refinement_iterations`, `search_iterations` variables exist but the runner may not loop based on these settings, executing a single pass or none.
- Condition/Gating Misconfiguration:
  - `model_deployment_prep` has a condition `${validation_and_debugging.output.final_performance_metrics.accuracy} > 0.85`. Similar implicit conditions may short-circuit other tasks if their inputs don’t exist.
  - If the runner interprets missing keys as falsy and still marks tasks complete, phases appear skipped.
- Stream-JSON Chaining Disabled/Incorrect:
  - The runtime may not actually pass outputs between agents unless `--chaining` or `stream-json` mode is active and correctly implemented for each task.
- Agent Type-to-Executor Mapping:
  - `type`: `optimizer`, `architect`, `tester` may not have implemented executors or default to a no-op that returns "completed" immediately.

## Requirements (for Coding Agent)
- Implement strict input validation on task start:
  - Fail the task with a clear error if required inputs are missing or empty.
  - Required inputs for refinement phases:
    - `ablation_analysis`: needs `foundation_building.output` and `initial_results`.
    - `targeted_refinement`: needs `highest_impact_component`, `impact_scores`, `improvement_opportunities`.
    - `ensemble_creation`: needs optimized components and performance improvements.
    - `validation_and_debugging`: needs ensemble model and full pipeline config.
- Enforce dependencies and failure policy:
  - If an upstream dependency is failed or incomplete, do not start downstream tasks.
  - Make `failurePolicy` configurable per task; default to `halt` for refinement phases.
- Honor iteration variables:
  - Use `variables.search_iterations` and `variables.refinement_iterations` to loop execution with logging of iteration indexes and outputs.
- Implement agent executors for `optimizer`, `architect`, `tester` types:
  - Map to concrete handlers that perform real work (code changes, hyperparameter search, ensemble assembly, CV runs).
  - Produce artifacts in `outputs/` and logs in `logs/` with task-specific subfolders.
- Strengthen stream-chaining:
  - When `--chaining` or `settings.auto_documentation/save_intermediate_results` is true, persist and pass structured outputs.
  - Validate the schema of passed data; log schema diffs and coercions.
- Add observability:
  - Per-iteration timing, artifact counts, and metrics.
  - Emit warnings if a task completes in <2s with no outputs.

## Taboos
- Do not silently mark tasks as completed when inputs are missing.
- Do not bypass iteration loops without emitting a warning/error explaining why.
- Do not change the MLE-STAR methodology or remove core phases.
- Do not introduce blocking external dependencies without fallbacks.

## Acceptance Criteria
- Refinement-related tasks show multi-iteration logs with durations consistent with actual work (not <10s trivial completions).
- Task outputs are persisted and consumed by next phases (validated via logs and generated files).
- Missing inputs cause tasks to fail with actionable error messages.
- Stream-chaining confirmed in logs; schemas validated and reported.
- Ensemble and validation phases produce artifacts (models, metrics, reports) in `./outputs/`.

## Implementation Plan (High-Level)
1. Task Engine Hardening
   - Add required-inputs validation to task runner.
   - Enforce dependency completion and configurable failure policy.
2. Iteration Support
   - Read iteration variables and loop; log per-iteration work, metrics.
3. Agent Executor Mapping
   - Implement handlers for `optimizer`, `architect`, `tester` with real actions and artifacts.
4. Stream-Chaining & Schema Validation
   - Ensure outputs are saved and typed; validate before consumption.
5. Observability
   - Add per-task timing, output counters, warnings for trivial completions.

## Files/Areas to Modify
- Workflow runner/orchestrator in `src/cli` and related automation code.
- Agent executors for non-foundation phases.
- Input resolution and templating layer handling `${task.output.*}`.
- Logging and persistence modules.

## Suggested Developer Tasks
- Add a `requiredInputs` array to tasks and enforce on start.
- Implement `iterationController` utility honoring `variables.refinement_iterations` and `variables.search_iterations`.
- Create executors:
  - `optimizer`: runs ablation + targeted refinement loop, writes `impact_scores.json`, `refinement_history.json`, `best_config.json`.
  - `architect`: builds ensembles, writes `ensemble_model.pkl`, `ensemble_report.json`.
  - `tester`: runs CV, leakage checks, writes `validation_report.json`, `metrics.json`.
- Add `--chaining` runtime flag detection; wire outputs between phases; persist to `outputs/<task-id>/`.
- Add guardrails: error when duration <2s and outputs are empty for non-trivial tasks.

## How to Validate After Fix
```bash
# Dry-run with verbose logging and chaining
claude-flow automation mle-star \
  --dataset "data/month_data_3y-train.csv" \
  --target MONTH_QTY \
  --output "./outputs/" \
  --name "Time Series Forecasting" \
  --chaining \
  --claude \
  --verbose

# Check logs for iteration counts and artifacts
ls -lah outputs/ablation_analysis/
ls -lah outputs/targeted_refinement/
ls -lah outputs/ensemble_creation/
ls -lah outputs/validation_and_debugging/

# Confirm non-trivial durations and outputs exist
```
